woocommerce/woocommerce
# Set in-app product card price meta dynamically

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:


Adds support for using the `currency`, `billingPeriod` and `billingPeriodInterval` properties returned by the `/featured` and `/search` endpoints on WooCommerce.com to set the price currency and suffix dynamically on product cards in the in-app marketplace.



Closes https://github.com/Automattic/woocommerce.com/pull/21635



### How to test the changes in this Pull Request:



#### Prep

**Note:** the related API changes are already merged into WCCOM, so you can also test without doing these prep steps first, but pointing to your local WCCOM will enable you to make changes to the products to test the different states.

To have your local WooCommerce core wp-env environment display products from your local WCCOM environment, you first need a running instance of WCCOM. 

1. Make sure your WCCOM environment is running on the latest `trunk`.
2. Make sure you have a couple subscription products with different billing periods, and one with a non-default billing interval (see https://github.com/Automattic/woocommerce.com/pull/21635)
3. Go to https://woocommerce.test/wp-admin/edit.php?post_type=featured_product and add the subscription products to a group. Also make sure you have some regular paid and free products added to the groups (each group displays up to 3 products in-app).
4. Check that the products are being returned from the featured endpoint https://woocommerce.test/wp-json/wccom-extensions/3.0/featured?country=US. (You'll probably need to flush the transient that caches this on your local WCCOM with `npm run wp:cli cache flush`.)

Next you'll need to point your WC Core environment to your local WCCOM:

1. Add a mu-plugin to your WC Core wp-env (as suggested in p6JqRr-aWb-p2), and put these two filters in it:

```php
add_filter( 'http_request_host_is_external', '__return_true' );

add_filter( 'http_request_args', function ( $r ) {
	$r['reject_unsafe_urls'] = false;
	$r['sslverify']          = false;
	$r['timeout']            = 60;

	return $r;
} );
```

2. Get the ID of wp-env's WordPress container.

<img width="500" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1647564/13509f82-962c-4f88-92ce-5f39222f2f64">

3. Link the container to woocommerce.test in the Docker network.

```bash
docker network connect woocommerce.test <container ID>
```

#### Test

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Check out this branch
2. Apply [this patch](https://github.com/user-attachments/files/16887916/wccom.patch) to point your in-app Marketplace to your local WCCOM
3. Go to `/plugins/woocommerce` and run `pnpm --filter='@woocommerce/plugin-woocommerce' build`
4. Visit the [in-app marketplace](http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fextensions). 
5. Check the products on the Discover tab. Confirm they're displaying the `$` currency symbol, and the correct suffix text ("annually" for a simple paid product, or the corresponding period and optionally interval for subscription products.
![Extensions_‹_WooCommerce_‹_WooCommerce_Core_E2E_Test_Suite](https://github.com/user-attachments/assets/55ddca26-c1ae-4519-ab7f-530304d1d970)

6. Switch to the `Browse` or `Themes` tab. Confirm the product prices there also show correctly. (If you don't see your subscription products, type a search term in the Search field that will return them.)
7. Edit this line: https://github.com/woocommerce/woocommerce/blob/34781007c477cf88cbcd498be9ab8fd2e84dd77f/plugins/woocommerce-admin/client/marketplace/components/product-card/product-card-footer.tsx#L89
Replace `product.currency` with a supported currency code, e.g. `EUR`
8. Refresh the Extensions page and confirm the price cards correctly show the `€` currency symbol (or the correct symbol for the code you used)



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 
Product cards in the in-app marketplace show currency and billing period data dynamically, based on data received from WooCommerce.com
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
